library/core/src/ptr/non_null.rs

```diff
@@ -430,6 +432,10 @@ impl<T: ?Sized> NonNull<T> {
     #[must_use = "this returns the result of the operation, \
                   without modifying the original"]
     #[inline]
+    //Address presaervation
+    #[ensures(|result: &NonNull<U>|  result.as_ptr() as *const () as usize == self.as_ptr() as *const () as usize )]
```

Comment:
Is the case to `*const ()` necessary, or can we can cast directly to `usize`?

Reply:
Yeah, I tried removing it and this errors comes:`
    = help: cast through a thin pointer first
`